doc/developers/developing_scikit.rst

```diff
@@ -0,0 +1,727 @@
+is not met, an exception of type ``ValueError`` should be raised.
+
+``y`` might be ignored in the case of unsupervised learning. However, to
+make it possible to use the estimator as part of a pipeline that can
+mix both supervised and unsupervised transformers, even unsupervised
+estimators need to accept a ``y=None`` keyword argument in
+the second position that is just ignored by the estimator.
+For the same reason, ``fit_predict``, ``fit_transform``, ``score``
+and ``partial_fit`` methods need to accept a ``y`` argument in
+the second place if they are implemented.
+
+The method should return the object (``self``). This pattern is useful
+to be able to implement quick one liners in an IPython session such as::
+
+  y_predicted = SVC(C=100).fit(X_train, y_train).predict(X_test)
+
+Depending on the nature of the algorithm, ``fit`` can sometimes also
+accept additional keywords arguments. However, any parameter that can
+have a value assigned prior to having access to the data should be an
+``__init__`` keyword argument. **fit parameters should be restricted
+to directly data dependent variables**. For instance a Gram matrix or
+an affinity matrix which are precomputed from the data matrix ``X`` are
+data dependent. A tolerance stopping criterion ``tol`` is not directly
+data dependent (although the optimal value according to some scoring
+function probably is).
+
+When ``fit`` is called, any previous call to ``fit`` should be ignored. In
+general, calling ``estimator.fit(X1)`` and then ``estimator.fit(X2)`` should
+be the same as only calling ``estimator.fit(X2)``. However, this may not be
+true in practice when ``fit`` depends on some random process, see
+:term:`random_state`. Another exception to this rule is when the
+hyper-parameter ``warm_start`` is set to ``True`` for estimators that
+support it. ``warm_start=True`` means that the previous state of the
+trainable parameters of the estimator are reused instead of using the
+default initialization strategy.
+
+Estimated Attributes
+^^^^^^^^^^^^^^^^^^^^
+
+Attributes that have been estimated from the data must always have a name
+ending with trailing underscore, for example the coefficients of
+some regression estimator would be stored in a ``coef_`` attribute after
+``fit`` has been called.
+
+The estimated attributes are expected to be overridden when you call ``fit``
+a second time.
+
+Optional Arguments
+^^^^^^^^^^^^^^^^^^
+
+In iterative algorithms, the number of iterations should be specified by
+an integer called ``n_iter``.
+
+Pairwise Attributes
+^^^^^^^^^^^^^^^^^^^
+
+An estimator that accept ``X`` of shape ``(n_samples, n_samples)`` and defines
+a :term:`_pairwise` property equal to ``True`` allows for cross-validation of
+the dataset, e.g. when ``X`` is a precomputed kernel matrix. Specifically,
+the :term:`_pairwise` property is used by ``utils.metaestimators._safe_split``
+to slice rows and columns.
+
+.. _testing_coverage:
+
+Testing and improving test coverage
```

Comment:
I'm now thinking this belongs back in the contributing guide. It's more specific to scikit-learn core than I had realised. (It would be nice to eventually include some general comments here on how to test estimators.)